torch/csrc/api/src/optim/optimizer.cpp

```diff
@@ -40,11 +129,35 @@ size_t OptimizerBase::size() const noexcept {
   return parameters_.size();
 }
 
+OptimizerOptions& OptimizerBase::defaults() noexcept {
+  return *defaults_.get();
+}
+
+const OptimizerOptions& OptimizerBase::defaults() const noexcept {
+  return *defaults_.get();
+}
+
+std::vector<OptimizerParamGroup>& OptimizerBase::param_groups() noexcept {
+  return param_groups_;
+}
+
+const std::vector<OptimizerParamGroup>& OptimizerBase::param_groups() const noexcept {
+  return param_groups_;
+}
+
+ska::flat_hash_map<std::string, std::unique_ptr<OptimizerParamState>>& OptimizerBase::state() noexcept {
+  return state_;
+}
+
+const ska::flat_hash_map<std::string, std::unique_ptr<OptimizerParamState>>& OptimizerBase::state() const noexcept {
+  return state_;
+}
+
 Tensor& OptimizerBase::buffer_at(std::vector<Tensor>& buffers, size_t index) {
   if (buffers.size() <= index) {
     buffers.reserve(index);
     for (auto i = buffers.size(); i <= index; ++i) {
-      buffers.push_back(torch::zeros_like(parameters_.at(i)));
+      buffers.emplace_back(torch::zeros_like(parameters_.at(i)));
```

Comment:
Is there a reason we should change to `emplace_back`?

Reply:
yeah emplace_back, instead of taking a value_type it takes a variadic list of arguments, and so you can perfectly forward the arguments and construct directly an object into a container without a temporary at all